core/src/main/scala/chisel3/experimental/hierarchy/InstantiateImpl.scala

```diff
@@ -88,9 +88,10 @@ private[chisel3] trait InstantiateImpl {
     args: K,
     f:    K => A
   ): Definition[A] = {
+    val modulePrefix = Builder.getModulePrefixList
     Builder.contextCache
       .getOrElseUpdate(
-        CacheKey(boxAllData(args), implicitly[ru.WeakTypeTag[A]]), {
+        CacheKey(boxAllData(args), implicitly[ru.WeakTypeTag[A]], modulePrefix), {
```

Comment:
we should think about -- does/should modulePrefix need to be part of the cache key for an ExtModule? It's hard to know if the ExtModule is going to override its desiredName with Module.prefix or not.

But maybe it doesn't really matter because its not like an extmodule in firrtl has much cost to it

Reply:
This is a great point and question, and unfortunately the decision on whether or not the modulePrefix has to be made before instantiating the module (we can't like, query the user-overridden `desiredName`).

Reply:
Because of how the diff is, I can't do this as a "suggestion", but I think lines 91-94 should basically be:
```scala
    val ttag = implicitly[ru.WeakTypeTag[A]]
    val modulePrefix = if (ttag.tpe <:< typeOf[Prefixable]) Builder.getModulePrefixList else Nil
    Builder.contextCache
      .getOrElseUpdate(
        CacheKey(boxAllData(args), ttag, modulePrefix), {
```
Again, with `Prefixable` as a strawman proposal for the opt-in trait, I'm open to better names.

Reply:
I retract my suggestion above--we should always make the prefix be part of the `CacheKey` even if it leads to some unnecessary duplication of definitions for ExtModules/BlackBoxes. The reasoning is that even if we added an opt-in trait like my proposed `Prefixable`--if a user didn't use `Prefixable` but override's `desiredName` to use the prefix, we have no way to detect that they did that, but if we didn't use the module prefix as part of the key, the `Instantiable` would do the wrong thing because it would hit in the cache for the same ExtModule despite the prefix being different. Thus we have to defensively assume users could always override desiredName to incorporate the prefix, so we must always factor it in here.